atom/browser/net/atom_network_delegate.cc

```diff
@@ -0,0 +1,388 @@
+// Copyright (c) 2015 GitHub, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "atom/browser/net/atom_network_delegate.h"
+
+#include "atom/common/native_mate_converters/net_converter.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/resource_request_info.h"
+#include "net/url_request/url_request.h"
+
+using content::BrowserThread;
+
+namespace atom {
+
+namespace {
+
+std::string ResourceTypeToString(content::ResourceType type) {
+  switch (type) {
+    case content::RESOURCE_TYPE_MAIN_FRAME:
+      return "main_frame";
+    case content::RESOURCE_TYPE_SUB_FRAME:
+      return "sub_frame";
+    case content::RESOURCE_TYPE_STYLESHEET:
+      return "stylesheet";
+    case content::RESOURCE_TYPE_SCRIPT:
+      return "script";
+    case content::RESOURCE_TYPE_IMAGE:
+      return "image";
+    case content::RESOURCE_TYPE_OBJECT:
+      return "object";
+    case content::RESOURCE_TYPE_XHR:
+      return "xmlhttprequest";
+    default:
+      return "other";
+  }
+}
+
+AtomNetworkDelegate::BlockingResponse RunListener(
+    const AtomNetworkDelegate::Listener& callback,
+    scoped_ptr<base::DictionaryValue> details) {
+  return callback.Run(*(details.get()));
+}
+
+bool MatchesFilterCondition(
+    net::URLRequest* request,
+    const AtomNetworkDelegate::ListenerInfo& info) {
+  if (!info.url_patterns.empty()) {
+    auto url = request->url();
+    for (auto& pattern : info.url_patterns)
+      if (pattern.MatchesURL(url))
+        return true;
+    return false;
+  }
+
+  return true;
+}
+
+scoped_ptr<base::DictionaryValue> ExtractRequestInfo(net::URLRequest* request) {
+  scoped_ptr<base::DictionaryValue> dict(new base::DictionaryValue());
+  dict->SetInteger("id", request->identifier());
+  dict->SetString("url", request->url().spec());
+  dict->SetString("method", request->method());
+  content::ResourceType resourceType = content::RESOURCE_TYPE_LAST_TYPE;
+  auto info = content::ResourceRequestInfo::ForRequest(request);
+  if (info)
+    resourceType = info->GetResourceType();
+  dict->SetString("resourceType", ResourceTypeToString(resourceType));
+  dict->SetDouble("timestamp", base::Time::Now().ToDoubleT() * 1000);
+
+  return dict.Pass();
+}
+
+base::DictionaryValue* GetRequestHeadersDict(
```

Comment:
The `base::DictionaryValue*` should also be wrapped in `scoped_ptr` here,  and in `GetResponseHeadersDict` bellow.